### PR TITLE
Insta cast

### PR DESCRIPTION
**Pull request by JonaLam**
# Description

Added card dragging and insta cast
small rework to effect base to allow for target all effects

## Related issue(s)
 #21 

## List of changes

-Added dragging cards when pressed
-Added cast type enum which automaticly sorts cards into either insta cast or target
-instacast cards can be played by pressing anywhere above player's cards
-target cards move to the side when in cast zone so that the player can target who they want
-Added enum to EffectBase which allows for damageing only a single enemy or all enemies, as well as having cards that do both.
-Right clicking unqueues card

### CODE REVIEW

**Comment by cheesefrycook** [comment]
Just pulled and tested, seems to work well! Will do my review in a bit

File: Cards/CardBase.gd
Comment (cheesefrycook) [inline Cards/CardBase.gd:82]: I think it will probably be better to move this logic into the Battler, then we could remove the `battler_reference` that we have here. However I believe Turtyo has mentioned there will be future changes to how targeting works. So for now this works fine I suppose.

File: Cards/CardContainer.gd
Comment (cheesefrycook) [inline Cards/CardContainer.gd:247]: The problem with having the right click in the ClickHandler is that you always have to right click the card to cancel it. This is a bit annoying when you have a SINGLE_TARGET type card, because it's not going to be at your mouse position

Comment (cheesefrycook) [reply Cards/CardContainer.gd:247]: So maybe we can check for the right click in `_process` using `Input.is_mouse_button_pressed` or something similar. This way it isn't checking for a click on a specific card

File: Cards/Movement State/MoveState_Queued.gd
Comment (cheesefrycook) [inline Cards/Movement State/MoveState_Queued.gd:39]: Maybe we should introduce a new `TARGETING` card state. So the idea is that when you place a card in the play area, if it's `SINGLE_TARGET`, then it enters the `TARGETING` state. Right now, the card will move to the `TARGET_CARD_POSITION` immediately when it enters the play area, which is a bit confusing to me.

Comment (cheesefrycook) [reply Cards/Movement State/MoveState_Queued.gd:39]: Actually playing with it a bit more, I think it's fine in the current state. I'd be curious to hear other people's thoughts though

File: Cards/CardBase.gd
Comment (JonaLam) [reply Cards/CardBase.gd:82]: I guess, but it feels more intuitve for me that playing cards happens in the card script, but we can see what's better if changes happen

File: Cards/CardContainer.gd
Comment (JonaLam) [reply Cards/CardContainer.gd:247]: Ye, that's fair, I'll do that

File: Cards/Movement State/MoveState_Queued.gd
Comment (JonaLam) [reply Cards/Movement State/MoveState_Queued.gd:39]: I think it will feel better if we add some more effects to make it less confusing, I think it's better that it doesn't require an extra click

**Comment by Turtyo** [comment]
Will check this tomorrow. i'm not too sure about "Added enum to EffectBase which allows for damageing only a single enemy or all enemies, as well as having cards that do both." for two reasons:
- This should be put in the general const file (which exists only in the map generation for now, but it's very useful so i might just do a small issue / PR for that only)
- i'll see how that plays with effects, because effects in themselves already have the ability to damage one or all (and you only need two of those effects to do both)

Maybe right click shouldn't unqueue card because it might be counter intuitive. We could unqueue the card if we click in some areas instead (hand, areas without enemies...). i believe the hovering / selecting a card already works in that way also.

File: Cards/Movement State/MoveState_Queued.gd
Comment (cheesefrycook) [reply Cards/Movement State/MoveState_Queued.gd:39]: Yes I think effects will help a lot. I tested in StS as well and it works pretty much exactly like you have it here

File: Cards/CardBase.gd
Comment (Turtyo) [inline Cards/CardBase.gd:38]: `EffectBase` already has an `apply_effect` function, and this is not doing much, so I don't really see the goal of this function

File: Cards/CardBase.gd
Comment (Turtyo) [inline Cards/CardBase.gd:45]: this honestly feels like it could be moved in the `_apply_all_effects` function, because this is basically what it does

File: Cards/CardBase.gd
Comment (Turtyo) [inline Cards/CardBase.gd:52]: why use `push_front` instead of `append` ? this will make effects play in reverse
for example if you have a card that says "Does A, then B, then C", when we read it the list will be `[A]` then we push B to the front it becomes `[B,A]` then `[C,B,A]`

File: Cards/CardBase.gd
Comment (Turtyo) [reply Cards/CardBase.gd:74]: I agree that matching the type should probably be moved to the battler and given to the card_play maybe. And yes, I will open an issue for the targeting i'm thinking of to see if we agree.
The problem of having all the logic by inside `on_card_play` is also that we will need to reformat it every time we want to change something about the logic. If instead we use functions that do things, we only need to change the sub-function without needing to change the card play in itself. `on_card_play` should only deal with having to play the card, and not knowing how to build the list of effects or knowing how to build the list of targets. Those things should be dealt in their respective functions, and `on_card_play` should only call them.

File: Cards/CardContainer.gd
Comment (Turtyo) [inline Cards/CardContainer.gd:427]: what is the 2 here ?

File: Cards/CardContainer.gd
Comment (Turtyo) [inline Cards/CardContainer.gd:416]: wasn't all the queue-ing already done in functions that Cheesy made ?

**Review by Turtyo — COMMENTED** [review]
I think some functions need to be broken down a bit to be more modular, and so that we don't need to modify their logic every time we change a little thing, and instead change the logic of the sub-functions

**Comment by Turtyo** [comment]
@JonaLam will let you solve the merge conflict, i checked it's just the testing scene, shouldn't be hard

File: Cards/CardBase.gd
Comment (JonaLam) [reply Cards/CardBase.gd:38]: That's fair, I'll change it

File: Cards/CardBase.gd
Comment (JonaLam) [reply Cards/CardBase.gd:52]: I don't know why I did it like that either, I'll change it

File: Cards/CardBase.gd
Comment (JonaLam) [reply Cards/CardBase.gd:74]: That's a good point, I'll at least move some of the logic out to other functions

File: Cards/CardContainer.gd
Comment (JonaLam) [reply Cards/CardContainer.gd:416]: I don't think so, I couldn't find anything like that, and the on click function didn't call anything like that so I don't know

File: Cards/CardContainer.gd
Comment (JonaLam) [reply Cards/CardContainer.gd:427]: It's for right clicking to cancel, I'll comment it at least for readability

Comment (Turtyo) [reply Cards/CardContainer.gd:429]: using a variable name would be even better

File: Cards/CardBase.gd
Comment (Turtyo) [inline Cards/CardBase.gd:54]: doesn't this assume all the effects have the same targets ? this might not be the case for a card like "does 1 damage to a selected enemy, heal 1 to self"

File: Cards/CardBase.gd
Comment (Turtyo) [inline Cards/CardBase.gd:67]: the name is a bit ambiguous. I would expect it to mean "all the effects of the card" but you mean it as "all the effects that apply to a target ALL"
`get_effects_target_all` would probably be a clearer name

File: Cards/CardContainer.gd
Comment (Turtyo) [reply Cards/CardContainer.gd:416]: @cheesefrycook can you confirm ? I thought I saw something like that in your PR ? or was it a discussion we had but chose to deal with later ?

File: Cards/CardBase.gd
Comment (JonaLam) [reply Cards/CardBase.gd:54]: Right now, they do all have the same target, and it's kinda difficult to affect both sides with one card. There should definitaly be some changes to this, but maybe it should happen with the target changes

File: Cards/CardBase.gd
Comment (JonaLam) [reply Cards/CardBase.gd:67]: I can change that

File: Cards/CardBase.gd
Comment (Turtyo) [reply Cards/CardBase.gd:54]: did you check #42 in link with this ?

Comment (JonaLam) [reply Cards/CardBase.gd:54]: Yeah, I saw, I was just thinking that fixing this issue would be better to do in #42, but I could do a bit of refactoring to enable the changes more

File: Cards/CardContainer.gd
Comment (cheesefrycook) [reply Cards/CardContainer.gd:416]: Unqueuing a card previously you had to click the card, so I handle that behavior in _on_card_clicked. So this function is useful if you want to unqueue without clicking I guess.

File: Cards/CardBase.gd
Comment (Turtyo) [reply Cards/CardBase.gd:54]: thank you

Comment (JonaLam) [reply Cards/CardBase.gd:54]: I can't actually think of a good way of fixing this without just refactoring the whole thing, so I think I'll leave it to #42

Comment (Turtyo) [reply Cards/CardBase.gd:54]: alright fair enough, this might be a bit out of the scope anyway

**Comment by Turtyo** [comment]
@JonaLam don't forget to solve the merge conflict with the main

**Review by cheesefrycook — DISMISSED** [review]

File: Cards/CardBase.gd
Comment (Turtyo) [reply Cards/CardBase.gd:67]: @JonaLam unresolving this because i see it's back to the name I had a problem with

File: Cards/CardBase.gd
Comment (Turtyo) [inline Cards/CardBase.gd:46]: would rename to `effects_targeting_all`

File: Cards/CardBase.gd
Comment (Turtyo) [inline Cards/CardBase.gd:58]: would rename to `get_effects_targeting_single`

File: Cards/CardContainer.gd
Comment (Turtyo) [inline Cards/CardContainer.gd:318]: maybe naming it `list_targets` would be less confusing

File: Cards/Resource/Card_DamageAll.tres
Comment (Turtyo) [inline Cards/Resource/Card_DamageAll.tres:8]: why set the card title to null ?

File: Cards/Resource/Card_DamageAll.tres
Comment (Turtyo) [inline Cards/Resource/Card_DamageAll.tres:9]: missing description

File: Cards/Resource/Card_DamageAll.tres
Comment (Turtyo) [inline Cards/Resource/Card_DamageAll.tres:10]: This looks like the array is empty, meaning the card doesn't do anything

File: Cards/Resource/Card_Damage_EVERYTHING.tres
Comment (Turtyo) [inline Cards/Resource/Card_Damage_EVERYTHING.tres:19]: missing `card_title`

**Review by Turtyo — CHANGES_REQUESTED** [review]

File: #Scenes/TestingScene.tscn
Comment (Turtyo) [inline #Scenes/TestingScene.tscn:155]: why was this all removed btw ?

File: Cards/CardBase.gd
Comment (Turtyo) [reply Cards/CardBase.gd:43]: After reading again, it does more than just `_apply_all_effects, so it's fine

File: Cards/CardBase.gd
Comment (Turtyo) [reply Cards/CardBase.gd:55]: Could you give us a little summary of what you moved out ? I don't remember how it was looking at that point

File: Cards/CardContainer.gd
Comment (JonaLam) [reply Cards/CardContainer.gd:428]: Do we have a file like that?

Comment (Turtyo) [reply Cards/CardContainer.gd:428]: I don't think we do. My PR has a `GLOBAL_VAR.gd` file but that's a bit different.
Basically you go to project settings, autoload and you create a new node file from there then populate it with the variables you need. when you want to call it you do `your_file_class_name.variable`

Comment (JonaLam) [reply Cards/CardContainer.gd:428]: Do you think it would be better to just put it on the input map instead

Comment (Turtyo) [reply Cards/CardContainer.gd:428]: the input map ? i don't remember a file named like that

Comment (JonaLam) [reply Cards/CardContainer.gd:428]: it's a in the prodject settings and I think it would do about the same as the sulution you were proposing, just built into godot

Comment (Turtyo) [reply Cards/CardContainer.gd]: I see it allows reconfiguration also. This is a good idea !

**Review by Turtyo — CHANGES_REQUESTED** [review]
`test_health.gd` must be updated to use `Array[Entity]` instead of just `Entity` when using `on_card_play`
The default deck should also use a heal card to test that the insta-cast works on self also (I think it does given Damage EVERYTHING but I prefer to be sure)

File: Cards/CardBase.gd
Comment (JonaLam) [reply Cards/CardBase.gd:55]: I moved out the filtering of single and all target attacks into get_effects_targeting_single and get_effects_target_all respectively, as well as how the function got the array of every target into get_all_targets in the battle manager

**Review by cheesefrycook — DISMISSED** [review]

**Review by Turtyo — CHANGES_REQUESTED** [review]
Those 3 new commits don't address the initial issue that I reported ie all the `on_card_play` calls in `test_health.gd` must be updated to be given an `Array[Entity]` instead of just an `Entity`, given that the insta-cast works with a list of entities.

Do you not run the tests in local and check the stack trace ? This would allow you to find the errors easily. **Please run the GUT tests in local in the future.** 

I need to check why they are passing with the Github action, but this should be more of a safety net than the main way to run the tests anyway.

File: Cards/CardBase.gd
Comment (Turtyo) [reply Cards/CardBase.gd:55]: alright seems nice enough
changes will be made probably due to the targeting PR, but this looks okay
@cheesefrycook feel free to resolve if you feel like it's ok

**Review by cheesefrycook — APPROVED** [review]

File: Cards/CardContainer.gd
Comment (Tysterman74) [inline Cards/CardContainer.gd:429]: We're calling this in one spot, which is fine for now, but if any other file is going to utilize this action, we would want to create a global file to hold this in, so we can update it in one spot.

**Review by Tysterman74 — CHANGES_REQUESTED** [review]

File: Cards/CardContainer.gd
Comment (Turtyo) [reply Cards/CardContainer.gd:429]: I think it's a bit different because of the input map which already holds this

Comment (JonaLam) [reply Cards/CardContainer.gd:429]: I'm a bit confused, what is it that you want changed here?

Comment (Turtyo) [reply Cards/CardContainer.gd:428]: the fact you used a string "cancel_queued_card_in_battle" instead of using a variable that holds the string
but in fact, the string is linked to a variable (kinda) with the input map, meaning we already have a central place to hold changes to multiple instances if needed
the only problem we could have is if we decide we don't want to use the name "cancel_queued_card_in_battle", but technically that would be equivalent to changing the name of the variable in itself if we used one (so same problem).

At least if I understood correctly the input map

Comment (JonaLam) [reply Cards/CardContainer.gd:428]: So you think it's fine as is?

**Review by Turtyo — APPROVED** [review]

File: Cards/CardContainer.gd
Comment (Tysterman74) [reply Cards/CardContainer.gd:428]: I think this is fine for now, but if we see this being used in another file in another PR we should move it out to a global file.

**Review by Tysterman74 — APPROVED** [review]

**Pull request merged**
